Add homework2

HorutskaL opened the pull request.

HorutskaL committed:
Add homework2

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/beans/BeanB.java:
```diff
@@ -0,0 +1,20 @@
+        super(name, value);
+    }
+
+    private void customInitMethod() {
```
please update access modifiers for all init or destroy methods to `public`

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/beans/MyBean.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public void validate() {
```
That's very good that you've moved common validation logic to separate class, good job!

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/beans/MyValidator.java:
```diff
@@ -0,0 +1,5 @@
+package com.epam.spring.homework2.beans;
+
+public interface MyValidator {
```
Since this interface basically means - "whoever will implement me should validate itself"
The `Validatable` name would better fit here, please rename it

for ex. in java we have a lot of such interfaces(Comparable, Serializable etc)

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/config/AppConfig.java:
```diff
@@ -0,0 +1,47 @@
+    private int beanCValue;
+
+    @Bean(initMethod = "customInitMethod", destroyMethod = "customDestroyMethod")
+    @Order(1)
```
The `Order `annotation is not really appropriate here, since it allows to define absolute order between all the beans.
And what we need here accordingly to homework, is to define relative order among them

Try to update it using annotation `DependsOn `for 2 beans:

BeanB -> depends on D
BeanC -> depends on B

And see if these beans created in correct order

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/beans/MyBeanPostProcessor.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.springframework.beans.factory.config.BeanPostProcessor;
+
+public class MyBeanPostProcessor implements BeanPostProcessor {
```
This bean is missing `Component `annotation, that's why validation is not working, please add it and see if validation started to execute

tvladyka-photobox commented on src/main/java/com/epam/spring/homework2/config/AppConfigAnother.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Bean
+    public BeanA beanA() {
+        return new BeanA(beanAName, beanAValue);
```
Try to add some invalid value here(for ex `-1`) to see if validation works

HorutskaL committed:
Errors fixed. Suggestions taken into account.

HorutskaL committed:
Added invalid value.

HorutskaL commented:
@tvladyka-photobox I fixed errors and took your recommendations into account:
1) Updated access modifiers for all init or destroy methods to public;
2) Renamed interface;
3) Replaced Order annotation on DependsOn;
4) Added invalid value;
5) Added Component annotationt to MyBeanPostProcessor.

tvladyka-photobox commented:
Good job, PR approved
